File: Python/menstrualcycle.py
from datetime import datetime, timedelta

input = "%d/%m/%Y"
output = "%d %B, %Y"

def get_next_cycle_date(date, cycle_days):
	next = datetime.strptime(date, input).date() + timedelta(days=cycle_days)
	return next

def get_ovulation_date(date, cycle_days):
	od = get_next_cycle_date(date, cycle_days) - timedelta(days=14)
	return od

def get_fertile_window(date, cycle_days):
	fwb = get_ovulation_date(date, cycle_days) - timedelta(days=5)
	fwa = get_ovulation_date(date, cycle_days) + timedelta(days=1)
	return [fwb,fwa]

def get_safe_period(date, cycle_days):
	last_cycle = datetime.strptime(date, input)
	safe_period = get_fertile_window(date, cycle_days)
	next_cycle = get_next_cycle_date(date, cycle_days)
	return [[last_cycle, safe_period[0] - timedelta(days=1)],[safe_period[1] + timedelta(days=1), next_cycle]]
	
	
	
	
'''lcd = datetime.strptime("25/05/2025", input)
fertile_window = get_fertile_window("25/05/2025", 28)
safe_period = get_safe_period("25/05/2025", 28)

print("Last cycle date: " + lcd.strftime(output))

print("Expected next cycle: " + get_next_cycle_date("25/05/2025", 28).strftime(output))

print("Expected ovulation: " + get_ovulation_date("25/05/2025", 28).strftime(output))

print("Expected fertile window:" + fertile_window[0].strftime(output) + "to", fertile_window[1].strftime(output))

print("Expected safe period: " + safe_period[0][0].strftime(output) + " to ", safe_period[0][1].strftime(output) + " and " + safe_period[1][0].strftime(output) + " to " + safe_period[1][1].strftime(output))'''

File: Java/MenstrualCycle.java
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Arrays;

public class MenstrualCycle {
	static DateTimeFormatter fmt = DateTimeFormatter.ofPattern("dd/MM/yyyy");
	
	public static LocalDate getNextCycleDate(String date, int days) {
		LocalDate next = LocalDate.parse(date, fmt).plusDays(days);
		return next;
	}
	
	public static LocalDate getOvulationDate(String date, int days) {
		LocalDate od = getNextCycleDate(date, days).minusDays(14);
		return od;
	}
	
	public static LocalDate[] getFertileWindow(String date, int days) {
		LocalDate fwb = getOvulationDate(date, days).minusDays(5);
		LocalDate fwa = getOvulationDate(date, days).plusDays(1);
		return new LocalDate[]{fwb,fwa};
	}
	
	public static LocalDate[][] getSafePeriod(String date, int days) {
		LocalDate lastCycle = LocalDate.parse(date, fmt);
		LocalDate[] safePeriod = getFertileWindow(date, days);
		LocalDate nextCycle = getNextCycleDate(date, days);
		return new LocalDate[][] {{lastCycle, safePeriod[0].minusDays(1)},{safePeriod[1].plusDays(1), nextCycle}};
	}
	
	
	
	
	
	
	public static void main(String[] args) {
		DateTimeFormatter output = DateTimeFormatter.ofPattern("d MMMM, y");
		String date = "25/05/2025";
		LocalDate lcd = LocalDate.parse(date, fmt);
			
		System.out.println("Last cycle date: " + lcd.format(output));
		System.out.println("Expected next cycle: " + getNextCycleDate("25/05/2025", 28).format(output));
		System.out.println("Expected ovulation: " + getOvulationDate("25/05/2025", 28).format(output));
		
		
		System.out.println("Expected fertile window: " + getFertileWindow("25/05/2025", 28)[0].format(output) + " to " + getFertileWindow("25/05/2025", 28)[1].format(output));
		
		System.out.println("Expected safe period: " + getSafePeriod("25/05/2025", 28)[0][0].format(output) + " to " + getSafePeriod("25/05/2025", 28)[0][1].format(output) + " and " + getSafePeriod("25/05/2025", 28)[1][0].format(output) + " to " + getSafePeriod("25/05/2025", 28)[1][1].format(output));
	}
	
	
}